README.md
Infiniscroll
============

Infiniscroll is a jQuery plugin for infinite scrolling (also known as lazy loading, endless scrolling, autopager, endless pages etc.)
where new content is loaded via AJAX within the current page or content area as you scroll down. This is the same technique that
Twitter and Facebook use for their news feeds.

Usage
-----

Include the scripts in your HTML head:

```html
<script src="//ajax.googleapis.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>
<script src="infiniscroll.min.js"></script>
```

Set up your HTML something like the following (class names/IDs can be changed and passed into `infiniscroll()`):

```html
<div class="content">
	<h1>My Content</h1>
	<p>Lorem ipsum dolor sit amet...</p>

	<div class="pagination">
		...
		<a href="/page/2" class="next">Next</a>
	</div>
</div>
<div class="loading">Loading...</div>
```

Then simply hook up Infiniscroll:

```html
<script>
$(document).ready(function(){
	$('.content').infiniscroll();
});
</script>
```

Options
-------

Below is a list of all the options that you can pass into Infiniscroll and a description of what they do:

```javascript
navSelector: '.pagination',				// Selector for your static naivgation (this will be hidden)
nextSelector: '.pagination a.next',		// Selector for the NEXT link (e.g. to page 2)
loadingSelector: '.loading',			// Selector for the loading element
pageFragment: '.content',				// Selector for the content you want to extract from the response
scrollBuffer: 200,						// The scroll amount in px before the bottom of the page that Infiniscroll should start to load the next page
scrollOnLoad: true,						// Should the window scroll to the position of the newly loaded content (if the user is at the bottom of the page)
scrollOnLoadDistance: 200,				// The distance to scroll down when new content is loaded,
scrollOnLoadSpeed: 500,					// The speed to scroll down when new content is loaded
onInit: function(){},					// Callback after plugin has loaded
beforeContentLoaded: function(link){},	// Callback before new content is loaded
afterContentLoaded: function(html){}	// Callback after new content has been loaded
```

How?
----

There is a little known feature in the `.load()` method that lets you specify the CSS selector of the html you want to include.
jQuery will load in any local URL, then parse the html and grab only the elements you’ve defined with your selector. For example:

```javascript
// Load all post divs from page 2 into an off-DOM div
$('.content').load('/page/2/ .content',function(){
	$(this).appendTo('.content'); // once they're loaded, append them to our content area
});
```

Why?
----

See [http://gilbert.pellegrom.me/jquery-infiniscroll-plugin](http://gilbert.pellegrom.me/jquery-infiniscroll-plugin)

Credits
-------

Infiniscroll was created by [Gilbert Pellegrom](http://gilbert.pellegrom.me) from [Dev7studios](http://dev7studios.com). Released under the MIT license.

changelog.txt
*** Infiniscroll Changelog ***

2014.02.03 - version 1.0
 * Initial release

demo/more.php
<?php
sleep(2); // Simulate loading time...
$page = isset($_GET['page']) && is_numeric($_GET['page']) ? $_GET['page'] : 1;
?>
<!DOCTYPE html>
<html lang="en" class="no-js">
<head>
    <meta charset="utf-8" />
    <title>Infiniscroll Demo</title>
    <link rel="stylesheet" href="style.css" type="text/css" media="screen" />
</head>
<body>

    <div class="content">
        <h1>Page <?php echo $page; ?></h1>
        <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a justo rhoncus, convallis odio nec, tristique magna. Phasellus pharetra, augue eu ultrices semper, ipsum turpis venenatis neque, eget tincidunt nisi enim eu ante. Sed adipiscing, mauris vel cursus convallis, sapien augue imperdiet metus, a gravida diam quam venenatis lorem. Aenean malesuada, mauris sit amet faucibus convallis, odio dui sodales mi, vitae molestie risus ligula ac orci. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Donec varius ligula eu condimentum pharetra. Nullam nec interdum augue. Duis sit amet imperdiet turpis. Curabitur a mi eget arcu luctus tristique a ac arcu. Etiam bibendum enim id nisi fermentum, at rhoncus ipsum iaculis. Mauris posuere egestas porttitor. Mauris sed tortor vestibulum, gravida ligula quis, aliquet ligula. Vivamus vitae mattis lacus, sed consectetur neque.</p>
        <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Interdum et malesuada fames ac ante ipsum primis in faucibus. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Cras tincidunt sem ut tortor egestas imperdiet. Donec sem enim, mollis non eros in, rhoncus blandit nisl. Vestibulum orci turpis, congue vel turpis convallis, dictum eleifend velit. Duis mollis dui magna, eu aliquam purus vulputate vitae. Proin ornare bibendum congue. Nam venenatis id neque eget eleifend.</p>
        <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a justo rhoncus, convallis odio nec, tristique magna. Phasellus pharetra, augue eu ultrices semper, ipsum turpis venenatis neque, eget tincidunt nisi enim eu ante. Sed adipiscing, mauris vel cursus convallis, sapien augue imperdiet metus, a gravida diam quam venenatis lorem. Aenean malesuada, mauris sit amet faucibus convallis, odio dui sodales mi, vitae molestie risus ligula ac orci. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Donec varius ligula eu condimentum pharetra. Nullam nec interdum augue. Duis sit amet imperdiet turpis. Curabitur a mi eget arcu luctus tristique a ac arcu. Etiam bibendum enim id nisi fermentum, at rhoncus ipsum iaculis. Mauris posuere egestas porttitor. Mauris sed tortor vestibulum, gravida ligula quis, aliquet ligula. Vivamus vitae mattis lacus, sed consectetur neque.</p>
        <div class="pagination">
            <a href="more.php?page=<?php echo $page + 1; ?>" class="next">Next</a>
        </div>
    </div>

</body>
</html>

infiniscroll.js
/*
 * Infiniscroll v1.0
 * https://github.com/gilbitron/Infiniscroll
 *
 * Copyright 2014, Gilbert Pellegrom
 * Free to use and abuse under the MIT license.
 * http://opensource.org/licenses/MIT
 */

;(function($, window, document, undefined){

	var pluginName = 'infiniscroll',
		defaults = {
			navSelector: '.pagination',				// Selector for your static naivgation (this will be hidden)
			nextSelector: '.pagination a.next',		// Selector for the NEXT link (e.g. to page 2)
			loadingSelector: '.loading',			// Selector for the loading element
			pageFragment: '.content',				// Selector for the content you want to extract from the response
			scrollBuffer: 200,						// The scroll amount in px before the bottom of the page that Infiniscroll should start to load the next page
			scrollOnLoad: true,						// Should the window scroll to the position of the newly loaded content (if the user is at the bottom of the page)
			scrollOnLoadDistance: 200,				// The distance to scroll down when new content is loaded,
			scrollOnLoadSpeed: 500,					// The speed to scroll down when new content is loaded
			onInit: function(){},					// Callback after plugin has loaded
			beforeContentLoaded: function(link){},	// Callback before new content is loaded
			afterContentLoaded: function(html){}	// Callback after new content has been loaded
		};

	function Infiniscroll(element, options){
		this.el = element;
		this.$el = $(this.el);

		this.options = $.extend({}, defaults, options);

		this._isLoading = false;
		this._nextLink = null;

		this.init();
	}

	Infiniscroll.prototype = {

		init: function(){
			var $this = this;

			// Hide the pagination container
			$(this.options.navSelector).hide();
			// Setup the scroll event listener
			$(window).on('scroll', function(){
				$this.doScroll.apply($this);
			});

			this.options.onInit.call(this);
		},

		doScroll: function(){
			// Only load one page at a time
			if(this._isLoading) return;

			// Do we need to load the next page?
			if($(window).scrollTop() >= ($(document).height() - $(window).height() - this.options.scrollBuffer)){
				this._isLoading = true;

				// Get the next page link
				if(!this._nextLink){
					this._nextLink = $(this.options.nextSelector);
				}

				if(this._nextLink.attr('href')){
					this.options.beforeContentLoaded.call(this, this._nextLink);
					$(this.options.loadingSelector).show();

					// Setup "iscroll" for use in the load() callback
					var iscroll = this,
						url = this._nextLink.attr('href');

					// Add pageFragment to URL if required
					if(this.options.pageFragment)
						url += ' '+ this.options.pageFragment;

					// Load the next page into an off-DOM div
					$('<div/>').load(url, function(){
						var html = $(this).children(),
							shouldScroll = false;

						// If user is waiting at bottom of page?
						if(iscroll.options.scrollOnLoad && $(window).scrollTop() === $(document).height() - $(window).height())
							shouldScroll = true;

						// Process new content
						html.find(iscroll.options.navSelector).hide();
						iscroll._nextLink = html.find(iscroll.options.nextSelector);

						// Hide the loader
						$(iscroll.options.loadingSelector).hide();
						// Actually add our new content to the container
						html.appendTo(iscroll.el);

						// Scroll to the new content
						if(shouldScroll){
							$('html, body').animate({
								scrollTop: $(window).scrollTop() + iscroll.options.scrollOnLoadDistance
							}, iscroll.options.scrollOnLoadSpeed);
						}

						iscroll._isLoading = false;
						iscroll.options.afterContentLoaded.call(iscroll, html);
					});
				}
			}
		}

	};

	$.fn[pluginName] = function(options){
		return this.each(function(){
			if(!$.data(this, pluginName)){
				$.data(this, pluginName, new Infiniscroll(this, options));
			}
		});
	};

})(jQuery, window, document);